File: app/src/main/java/de/kuschku/libquassel/syncables/types/impl/BufferSyncer.java
/*
 * QuasselDroid - Quassel client for Android
 * Copyright (C) 2016 Janne Koschinski
 * Copyright (C) 2016 Ken Børge Viktil
 * Copyright (C) 2016 Magnus Fjell
 * Copyright (C) 2016 Martin Sandsmark <martin.sandsmark@kde.org>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.kuschku.libquassel.syncables.types.impl;

import android.support.annotation.NonNull;
import android.util.SparseArray;
import android.util.SparseIntArray;

import java.util.HashMap;
import java.util.Map;

import de.kuschku.libquassel.BusProvider;
import de.kuschku.libquassel.client.Client;
import de.kuschku.libquassel.message.Message;
import de.kuschku.libquassel.primitives.types.QVariant;
import de.kuschku.libquassel.syncables.serializers.BufferSyncerSerializer;
import de.kuschku.libquassel.syncables.types.abstracts.ABufferSyncer;
import de.kuschku.libquassel.syncables.types.interfaces.QBacklogManager;
import de.kuschku.libquassel.syncables.types.interfaces.QBufferSyncer;
import de.kuschku.libquassel.syncables.types.interfaces.QBufferViewConfig;
import de.kuschku.util.observables.lists.ObservableComparableSortedList;
import de.kuschku.util.observables.lists.ObservableElement;
import de.kuschku.util.observables.lists.ObservableSet;
import de.kuschku.util.observables.lists.ObservableSortedList;

import static de.kuschku.util.AndroidAssert.assertNotNull;

public class BufferSyncer extends ABufferSyncer {

    @NonNull
    private final SparseArray<ObservableElement<Integer>> activities = new SparseArray<>();
    @NonNull
    private SparseIntArray lastSeenMsgs = new SparseIntArray();
    @NonNull
    private SparseIntArray markerLines = new SparseIntArray();
    @NonNull
    private Map<Integer, ObservableSet<Message.Type>> filters = new HashMap<>();

    public BufferSyncer(@NonNull Map<Integer, Integer> lastSeenMsgs, @NonNull Map<Integer, Integer> markerLines) {
        assertNotNull(lastSeenMsgs);
        assertNotNull(markerLines);

        for (int bufferId : lastSeenMsgs.keySet()) {
            this.lastSeenMsgs.put(bufferId, lastSeenMsgs.get(bufferId));
        }
        for (int bufferId : markerLines.keySet()) {
            this.markerLines.put(bufferId, markerLines.get(bufferId));
        }
    }

    @Override
    public int lastSeenMsg(int buffer) {
        return lastSeenMsgs.get(buffer, -1);
    }

    @Override
    public int markerLine(int buffer) {
        return markerLines.get(buffer, -1);
    }

    @Override
    public void _setLastSeenMsg(int buffer, int msgId) {
        assertNotNull(client);
        QBacklogManager backlogManager = client.backlogManager();
        assertNotNull(backlogManager);

        if (msgId < 0)
            return;

        int oldLastSeenMsg = lastSeenMsg(buffer);
        if (oldLastSeenMsg < msgId) {
            lastSeenMsgs.put(buffer, msgId);
        }
        setActivity(buffer, 0);
        ObservableComparableSortedList<Message> filtered = backlogManager.filtered(buffer);
        for (Message m : filtered) {
            addActivity(m);
        }
        client.bufferManager().bufferIds().notifyItemChanged(buffer);
        _update();
    }

    @Override
    public void _setMarkerLine(int buffer, int msgId) {
        if (msgId < 0)
            return;

        int oldMarkerline = markerLine(buffer);
        if (oldMarkerline < msgId) {
            markerLines.put(buffer, msgId);
            client.backlogStorage().setMarkerLine(buffer, msgId);
        }
        client.bufferManager().bufferIds().notifyItemChanged(buffer);
        _update();
    }

    @Override
    public void _requestSetLastSeenMsg(int buffer, int msgId) {
        // Do nothing, we’re on the client – the server will receive the sync just as expected
    }

    @Override
    public void _requestSetMarkerLine(int buffer, int msgId) {
        // Do nothing, we’re on the client – the server will receive the sync just as expected
    }

    @Override
    public void _requestRemoveBuffer(int buffer) {
        // Do nothing, we’re on the client – the server will receive the sync just as expected
    }

    @Override
    public void _removeBuffer(int buffer) {
        assertNotNull(client);

        for (QBufferViewConfig config : client.bufferViewManager().bufferViewConfigs()) {
            config.deleteBuffer(buffer);
        }
        this.filters.remove(buffer);
        markerLines.delete(buffer);
        lastSeenMsgs.delete(buffer);
        client.bufferManager().removeBuffer(buffer);
        _update();
    }

    @Override
    public void _requestRenameBuffer(int buffer, String newName) {
        // Do nothing, we’re on the client – the server will receive the sync just as expected
    }

    @Override
    public void _renameBuffer(int bufferId, @NonNull String newName) {
        assertNotNull(client);

        client.bufferManager().renameBuffer(bufferId, newName);
        _update();
    }

    @Override
    public void _requestMergeBuffersPermanently(int buffer1, int buffer2) {
        // Do nothing, we’re on the client – the server will receive the sync just as expected
    }

    @Override
    public void _mergeBuffersPermanently(int buffer1, int buffer2) {
        client.backlogStorage().merge(buffer1, buffer2);
        _removeBuffer(buffer2);
    }

    @Override
    public void _requestPurgeBufferIds() {
        // Do nothing, we’re on the client – the server will receive the sync just as expected
    }

    @Override
    public void _requestMarkBufferAsRead(int buffer) {
        assertNotNull(client);

        int lastMessage = client.backlogStorage().getLatest(buffer);
        if (lastMessage != -1) {
            requestSetLastSeenMsg(buffer, lastMessage);
        }
    }

    @Override
    public void _markBufferAsRead(int buffer) {
        assertNotNull(client);

        ObservableSortedList<Message> messages = client.backlogStorage().getUnfiltered(buffer);
        Message lastMessage = messages.last();
        if (messages.isEmpty() || lastMessage == null) {
            _setLastSeenMsg(buffer, -1);
            _setMarkerLine(buffer, -1);
        } else {
            _setLastSeenMsg(buffer, lastMessage.id);
            _setMarkerLine(buffer, lastMessage.id);
        }
    }

    @Override
    public void init(@NonNull String objectName, @NonNull BusProvider provider, @NonNull Client client) {
        super.init(objectName, provider, client);
        client.setBufferSyncer(this);
    }

    @Override
    public void _update(@NonNull Map<String, QVariant> from) {
        _update(BufferSyncerSerializer.get().fromLegacy(from));
    }

    @Override
    public void _update(@NonNull QBufferSyncer from) {
        lastSeenMsgs = from.lastSeenMsgs();
        markerLines = from.markerLines();
        _update();
    }

    public ObservableElement<Integer> activity(int bufferid) {
        assertNotNull(activities);
        ensureExistingActivity(bufferid);

        return activities.get(bufferid);
    }

    public void setActivity(int bufferid, int activity) {
        assertNotNull(activities);
        ensureExistingActivity(bufferid);

        activities.get(bufferid).set(activity);
    }

    private void ensureExistingActivity(int bufferid) {
        if (activities.get(bufferid) == null)
            activities.put(bufferid, new ObservableElement<>(0));
    }

    public void addActivity(int bufferid, int activity) {
        assertNotNull(activities);
        ensureExistingActivity(bufferid);

        activities.get(bufferid).set(activities.get(bufferid).get() | activity);
    }

    public void addActivity(int bufferid, @NonNull Message.Type type) {
        ensureExistingActivity(bufferid);
        addActivity(bufferid, type.value);
    }

    public void addActivity(@NonNull Message message) {
        int bufferId = message.bufferInfo.id;
        int lastSeenMsg = lastSeenMsg(bufferId);
        boolean filtered = getFilteredTypes(bufferId).contains(message.type);
        if (!filtered && message.id > lastSeenMsg) {
            addActivity(bufferId, message.type);
        }
    }

    @Override
    public SparseIntArray lastSeenMsgs() {
        return lastSeenMsgs;
    }

    @Override
    public SparseIntArray markerLines() {
        return markerLines;
    }

    @Override
    public ObservableSet<Message.Type> getFilteredTypes(int bufferId) {
        if (!this.filters.containsKey(bufferId)) {
            this.filters.put(bufferId, new ObservableSet<>());
            setFilters(bufferId, client.metaDataManager().hiddendata(client.coreId(), bufferId));
        }
        return this.filters.get(bufferId);
    }

    @Override
    public int getFilters(int bufferId) {
        int filters = 0x00000000;
        for (Message.Type type : getFilteredTypes(bufferId)) {
            filters |= type.value;
        }
        return filters;
    }

    @Override
    public void setFilters(int bufferId, int filters) {
        client.metaDataManager().setHiddendata(client.coreId(), bufferId, filters);
        for (Message.Type type : Message.Type.values()) {
            if ((filters & type.value) != 0) {
                getFilteredTypes(bufferId).add(type);
            } else {
                getFilteredTypes(bufferId).remove(type);
            }
        }
    }
}

File: lib/src/main/java/de/kuschku/libquassel/quassel/syncables/BufferSyncer.kt
/*
 * Quasseldroid - Quassel client for Android
 *
 * Copyright (c) 2020 Janne Mareike Koschinski
 * Copyright (c) 2020 The Quassel Project
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 3 as published
 * by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package de.kuschku.libquassel.quassel.syncables

import de.kuschku.libquassel.protocol.*
import de.kuschku.libquassel.protocol.Type
import de.kuschku.libquassel.quassel.BufferInfo
import de.kuschku.libquassel.quassel.syncables.interfaces.IBufferSyncer
import de.kuschku.libquassel.session.ISession
import de.kuschku.libquassel.session.NotificationManager
import de.kuschku.libquassel.util.Optional
import de.kuschku.libquassel.util.flag.minus
import de.kuschku.libquassel.util.irc.IrcCaseMappers
import io.reactivex.Observable
import io.reactivex.subjects.BehaviorSubject

class BufferSyncer constructor(
  var session: ISession,
  private val notificationManager: NotificationManager? = null
) : SyncableObject(session.proxy, "BufferSyncer"), IBufferSyncer {
  override fun deinit() {
    super.deinit()
    session = ISession.NULL
  }

  fun lastSeenMsg(buffer: BufferId): MsgId = _lastSeenMsg[buffer] ?: MsgId(0)
  fun liveLastSeenMsg(buffer: BufferId): Observable<MsgId> = live_lastSeenMsg.map {
    markerLine(buffer)
  }.distinctUntilChanged()

  fun liveLastSeenMsgs(): Observable<Map<BufferId, MsgId>> =
    live_lastSeenMsg.map { _lastSeenMsg.toMap() }

  fun markerLine(buffer: BufferId): MsgId = _markerLines[buffer] ?: MsgId(0)
  fun liveMarkerLine(buffer: BufferId): Observable<MsgId> =
    live_markerLines.map { markerLine(buffer) }.distinctUntilChanged()

  fun liveMarkerLines(): Observable<Map<BufferId, MsgId>> =
    live_markerLines.map { _markerLines.toMap() }

  fun activity(buffer: BufferId): Message_Types =
    _bufferActivities[buffer] ?: Message_Types.of()

  fun liveActivity(buffer: BufferId): Observable<Message_Types> =
    live_bufferActivities.map { activity(buffer) }.distinctUntilChanged()

  fun liveActivities(): Observable<Map<BufferId, Message_Types>> =
    live_bufferActivities.map { _bufferActivities.toMap() }

  fun highlightCount(buffer: BufferId): Int = _highlightCounts[buffer] ?: 0
  fun liveHighlightCount(buffer: BufferId): Observable<Int> =
    live_highlightCounts.map { highlightCount(buffer) }.distinctUntilChanged()

  fun liveHighlightCounts(): Observable<Map<BufferId, Int>> =
    live_highlightCounts.map { _highlightCounts.toMap() }

  fun bufferInfo(bufferId: BufferId) = _bufferInfos[bufferId]
  fun liveBufferInfo(bufferId: BufferId) =
    live_bufferInfos.map { bufferInfo(bufferId) }.distinctUntilChanged()

  fun bufferInfos(): Collection<BufferInfo> = _bufferInfos.values.toList()
  fun liveBufferInfos(): Observable<Map<BufferId, BufferInfo>> = live_bufferInfos.map { _bufferInfos.toMap() }

  override fun toVariantMap(): QVariantMap = mapOf(
    "Activities" to QVariant.of(initActivities(), Type.QVariantList),
    "HighlightCounts" to QVariant.of(initHighlightCounts(), Type.QVariantList),
    "LastSeenMsg" to QVariant.of(initLastSeenMsg(), Type.QVariantList),
    "MarkerLines" to QVariant.of(initMarkerLines(), Type.QVariantList)
  )

  override fun fromVariantMap(properties: QVariantMap) {
    initSetActivities(properties["Activities"].valueOr(::emptyList))
    initSetHighlightCounts(properties["HighlightCounts"].valueOr(::emptyList))
    initSetLastSeenMsg(properties["LastSeenMsg"].valueOr(::emptyList))
    initSetMarkerLines(properties["MarkerLines"].valueOr(::emptyList))
  }

  fun copy() = BufferSyncer(session).also {
    it.fromVariantMap(toVariantMap())
  }

  fun isEqual(other: BufferSyncer) =
    _bufferInfos == other._bufferInfos &&
    _lastSeenMsg == other._lastSeenMsg &&
    _markerLines == other._markerLines &&
    _bufferActivities == other._bufferActivities &&
    _highlightCounts == other._highlightCounts


  override fun initActivities(): QVariantList {
    val list: MutableList<QVariant_> = mutableListOf()
    for ((key, value) in _bufferActivities) {
      list.add(QVariant.of(key, QType.BufferId))
      list.add(QVariant.of(value.toInt(), Type.Int))
    }
    return list
  }

  override fun initHighlightCounts(): QVariantList {
    val list: MutableList<QVariant_> = mutableListOf()
    for ((key, value) in _highlightCounts) {
      list.add(QVariant.of(key, QType.BufferId))
      list.add(QVariant.of(value, Type.Int))
    }
    return list
  }

  override fun initLastSeenMsg(): QVariantList {
    val list: MutableList<QVariant_> = mutableListOf()
    for ((key, value) in _lastSeenMsg) {
      list.add(QVariant.of(key, QType.BufferId))
      list.add(QVariant.of(value, QType.MsgId))
    }
    return list
  }

  override fun initMarkerLines(): QVariantList {
    val list: MutableList<QVariant_> = mutableListOf()
    for ((key, value) in _markerLines) {
      list.add(QVariant.of(key, QType.BufferId))
      list.add(QVariant.of(value, QType.MsgId))
    }
    return list
  }

  override fun initSetActivities(data: QVariantList) {
    setActivities((0 until data.size step 2).map {
      Pair(
        data[it].value(BufferId(0)),
        Message_Type.of(data[it + 1].value(0))
      )
    })
  }

  fun setActivities(data: List<Pair<BufferId, Message_Types>>) {
    for ((buffer, activity) in data) {
      setBufferActivityInternal(buffer, activity)
    }
    live_bufferActivities.onNext(Unit)
  }

  override fun initSetHighlightCounts(data: QVariantList) {
    setHighlightCounts((0 until data.size step 2).map {
      Pair(
        data[it].value(BufferId(0)),
        data[it + 1].value(0)
      )
    })
  }

  fun setHighlightCounts(data: List<Pair<BufferId, Int>>) {
    for ((buffer, count) in data) {
      setHighlightCount(buffer, count)
    }
    live_highlightCounts.onNext(Unit)
  }

  override fun initSetLastSeenMsg(data: QVariantList) {
    setLastSeenMsg((0 until data.size step 2).map {
      Pair(
        data[it].value(BufferId(0)),
        data[it + 1].value(MsgId(0L))
      )
    })
  }

  fun setLastSeenMsg(data: List<Pair<BufferId, MsgId>>) {
    for ((buffer, msgId) in data) {
      setLastSeenMsg(buffer, msgId)
    }
    live_lastSeenMsg.onNext(Unit)
  }

  override fun initSetMarkerLines(data: QVariantList) {
    setMarkerLines((0 until data.size step 2).map {
      Pair(
        data[it].value(BufferId(0)),
        data[it + 1].value(MsgId(0L))
      )
    })
  }

  fun setMarkerLines(data: List<Pair<BufferId, MsgId>>) {
    for ((buffer, msgId) in data) {
      setMarkerLine(buffer, msgId)
    }
    live_markerLines.onNext(Unit)
  }

  fun initSetBufferInfos(infos: QVariantList?) {
    setBufferInfos(infos?.mapNotNull { it.value<BufferInfo>() }.orEmpty())
  }

  fun setBufferInfos(infos: List<BufferInfo>) {
    _bufferInfos.clear()
    for (info in infos) {
      _bufferInfos[info.bufferId] = info
    }
    live_bufferInfos.onNext(Unit)
  }

  override fun mergeBuffersPermanently(buffer1: BufferId, buffer2: BufferId) {
    removeBuffer(buffer2)
  }

  override fun removeBuffer(buffer: BufferId) {
    _lastSeenMsg.remove(buffer);live_lastSeenMsg.onNext(Unit)
    _markerLines.remove(buffer);live_markerLines.onNext(Unit)
    _bufferActivities.remove(buffer);live_bufferActivities.onNext(Unit)
    _highlightCounts.remove(buffer);live_highlightCounts.onNext(Unit)
    _bufferInfos.remove(buffer);live_bufferInfos.onNext(Unit)
    session.backlogManager.removeBuffer(buffer)
    notificationManager?.clear(buffer)
  }

  override fun renameBuffer(buffer: BufferId, newName: String?) {
    val bufferInfo = _bufferInfos[buffer]
    if (bufferInfo != null) {
      _bufferInfos[buffer] = bufferInfo.copy(bufferName = newName)
      live_bufferInfos.onNext(Unit)
    }
  }

  fun bufferInfoUpdated(info: BufferInfo) {
    val oldInfo = _bufferInfos[info.bufferId]
    if (info != oldInfo) {
      _bufferInfos[info.bufferId] = info
      live_bufferInfos.onNext(Unit)

      if (oldInfo == null) {
        session.bufferViewManager.handleBuffer(info, this)
      }
    }
  }

  override fun setLastSeenMsg(buffer: BufferId, msgId: MsgId) {
    if (msgId < MsgId(0))
      return

    val oldLastSeenMsg = lastSeenMsg(buffer)
    if (oldLastSeenMsg < msgId) {
      _lastSeenMsg[buffer] = msgId
      live_lastSeenMsg.onNext(Unit)
      super.setLastSeenMsg(buffer, msgId)
      notificationManager?.clear(buffer, msgId)
    }
  }

  override fun setMarkerLine(buffer: BufferId, msgId: MsgId) {
    if (msgId < MsgId(0) || markerLine(buffer) == msgId)
      return

    _markerLines[buffer] = msgId
    live_markerLines.onNext(Unit)
    super.setMarkerLine(buffer, msgId)
  }

  override fun setBufferActivity(buffer: BufferId, activity: Int) {
    setBufferActivity(buffer, Message_Type.of(activity))
  }

  fun setBufferActivityInternal(buffer: BufferId, activity: Message_Types) {
    super.setBufferActivity(buffer, activity.toInt())
    _bufferActivities[buffer] = activity
    live_bufferActivities.onNext(Unit)
  }

  fun setBufferActivity(buffer: BufferId, activity: Message_Types) {
    val oldActivity = activity(buffer)
    setBufferActivityInternal(buffer, activity)
    if ((activity - oldActivity).isNotEmpty()) {
      bufferInfo(buffer)?.let {
        session.bufferViewManager.handleBuffer(it, this, true)
      }
    }
  }

  override fun setHighlightCount(buffer: BufferId, count: Int) {
    super.setHighlightCount(buffer, count)
    _highlightCounts[buffer] = count
    live_highlightCounts.onNext(Unit)
  }

  fun all(
    bufferName: String? = null,
    bufferId: BufferId? = null,
    networkId: NetworkId? = null,
    type: Buffer_Types? = null,
    groupId: Int? = null
  ) = _bufferInfos.values.filter {
    bufferId == null || it.bufferId == bufferId
  }.filter {
    networkId == null || it.networkId == networkId
  }.filter {
    type == null || it.type == type
  }.filter {
    groupId == null || it.groupId == groupId
  }.filter {
    val caseMapper = IrcCaseMappers[session.networks[it.networkId]?.support("CASEMAPPING")]
    bufferName == null || caseMapper.equalsIgnoreCaseNullable(it.bufferName, bufferName)
  }

  fun liveAll(
    bufferName: String? = null,
    bufferId: BufferId? = null,
    networkId: NetworkId? = null,
    type: Buffer_Types? = null,
    groupId: Int? = null
  ) = liveBufferInfos().map {
    it.values.filter {
      bufferId == null || it.bufferId == bufferId
    }.filter {
      networkId == null || it.networkId == networkId
    }.filter {
      type == null || it.type == type
    }.filter {
      groupId == null || it.groupId == groupId
    }.filter {
      val caseMapper = IrcCaseMappers[session.networks[it.networkId]?.support("CASEMAPPING")]
      bufferName == null || caseMapper.equalsIgnoreCaseNullable(it.bufferName, bufferName)
    }
  }

  fun find(
    bufferName: String? = null,
    bufferId: BufferId? = null,
    networkId: NetworkId? = null,
    type: Buffer_Types? = null,
    groupId: Int? = null
  ) = all(bufferName, bufferId, networkId, type, groupId).firstOrNull()

  fun liveFind(
    bufferName: String? = null,
    bufferId: BufferId? = null,
    networkId: NetworkId? = null,
    type: Buffer_Types? = null,
    groupId: Int? = null
  ) = liveAll(bufferName, bufferId, networkId, type, groupId).map {
    Optional.ofNullable(it.firstOrNull())
  }

  override fun toString(): String {
    return "BufferSyncer(_lastSeenMsg=$_lastSeenMsg, _markerLines=$_markerLines, _bufferActivities=$_bufferActivities, _highlightCounts=$_highlightCounts, _bufferInfos=$_bufferInfos)"
  }

  private val _lastSeenMsg: MutableMap<BufferId, MsgId> = mutableMapOf()
  private val live_lastSeenMsg = BehaviorSubject.createDefault(Unit)

  private val _markerLines: MutableMap<BufferId, MsgId> = mutableMapOf()
  private val live_markerLines = BehaviorSubject.createDefault(Unit)

  private val _bufferActivities: MutableMap<BufferId, Message_Types> = mutableMapOf()
  private val live_bufferActivities = BehaviorSubject.createDefault(Unit)

  private val _highlightCounts: MutableMap<BufferId, Int> = mutableMapOf()
  private val live_highlightCounts = BehaviorSubject.createDefault(Unit)

  private val _bufferInfos = mutableMapOf<BufferId, BufferInfo>()
  private val live_bufferInfos = BehaviorSubject.createDefault(Unit)
}
